zerver/tests/test_auth_backends.py

```diff
@@ -850,3 +877,368 @@ def test_login_success_under_subdomains(self):
+        # type: () -> None
+        backend = self.backend
+        with self.settings(LDAP_APPEND_DOMAIN='zulip.com'):
+            username = backend.ldap_to_django_username('"hamlet@test"')
+            self.assertEqual(username, '"hamlet@test"@zulip.com')
+
+    def test_get_or_create_user_when_user_exists(self):
+        # type: () -> None
+        class _LDAPUser(object):
+            attrs = {'fn': ['Full Name'], 'sn': ['Short Name']}
+
+        backend = self.backend
+        email = 'hamlet@zulip.com'
+        user_profile, created = backend.get_or_create_user(email, _LDAPUser())
+        self.assertFalse(created)
+        self.assertEqual(user_profile.email, email)
+
+    def test_get_or_create_user_when_user_does_not_exist(self):
+        # type: () -> None
+        class _LDAPUser(object):
+            attrs = {'fn': ['Full Name'], 'sn': ['Short Name']}
+
+        ldap_user_attr_map = {'full_name': 'fn', 'short_name': 'sn'}
+
+        with self.settings(AUTH_LDAP_USER_ATTR_MAP=ldap_user_attr_map):
+            backend = self.backend
+            email = 'nonexisting@zulip.com'
+            user_profile, created = backend.get_or_create_user(email, _LDAPUser())
+            self.assertFalse(created)
+            self.assertEqual(user_profile.email, email)
+            self.assertEqual(user_profile.full_name, 'Full Name')
+
+    def test_get_or_create_user_when_realm_is_deactivated(self):
+        # type: () -> None
+        class _LDAPUser(object):
+            attrs = {'fn': ['Full Name'], 'sn': ['Short Name']}
+
+        ldap_user_attr_map = {'full_name': 'fn', 'short_name': 'sn'}
+
+        with self.settings(AUTH_LDAP_USER_ATTR_MAP=ldap_user_attr_map):
+            backend = self.backend
+            email = 'nonexisting@zulip.com'
+            realm = get_realm('zulip.com')
+            do_deactivate_realm(realm)
+            with self.assertRaisesRegexp(Exception, 'Realm has been deactivated'):
+                backend.get_or_create_user(email, _LDAPUser())
+
+    def test_django_to_ldap_username_when_domain_does_not_match(self):
+        # type: () -> None
+        backend = self.backend
+        email = 'hamlet@zulip.com'
+        with self.assertRaisesRegexp(Exception, 'Username does not match LDAP domain.'):
+            with self.settings(LDAP_APPEND_DOMAIN='acme.com'):
+                backend.django_to_ldap_username(email)
+
+    def test_login_failure_due_to_wrong_subdomain(self):
+        # type: () -> None
+        self.mock_ldap.directory = {
+            'uid=hamlet,ou=users,dc=zulip,dc=com': {
+                'userPassword': 'testing'
+            }
+        }
+        with self.settings(
+                REALMS_HAVE_SUBDOMAINS=True,
+                LDAP_APPEND_DOMAIN='zulip.com',
+                AUTH_LDAP_BIND_PASSWORD='',
+                AUTH_LDAP_USER_DN_TEMPLATE='uid=%(user)s,ou=users,dc=zulip,dc=com'):
+            user_profile = self.backend.authenticate('hamlet@zulip.com', 'testing', 'acme')
```

Comment:
please pass this as `realm_subdomain="acme"` ; it's best to pass kwargs as kwargs for clarity.  Same with the next few tests.